Throw an error if nbytes is called on a sparse tensor.

gchanan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#33897 Throw an error if nbytes is called on a sparse tensor.**

Differential Revision: [D20146388](https://our.internmc.facebook.com/intern/diff/D20146388)

gchanan committed:
Throw an error if nbytes is called on a sparse tensor.

[ghstack-poisoned]

zou3519 reviewed (commented):
does it matter that this is BC-breaking?

gchanan committed:
Update on "Throw an error if nbytes is called on a sparse tensor."


Differential Revision: [D20146388](https://our.internmc.facebook.com/intern/diff/D20146388)

[ghstack-poisoned]

gchanan commented:
I can mark it BC breaking, but it's apparently C++ only and I doubt people are doing sparse tensor-y things in C++.

gchanan commented on aten/src/ATen/templates/TensorBody.h:
```diff
@@ -227,6 +227,10 @@ class CAFFE2_API Tensor {
   // it reports the memory the tensor would take *if* it were contiguous.
   // Defined to be numel() * itemsize()
   size_t nbytes() const {
+    AT_CHECK(layout () != at::kSparse,
```
is this used anywhere?  Make it sure it's fast if it's used in someplace important.

gchanan replied:
I checked and it's used in two places:
1) In the quantized constructor, because they don't use normal copy for some reason (use memcpy instead).
2) In the java bindings to create a byte buffer.

Since neither of these need fast nbytes, we are going to commit this.

gchanan committed:
Update on "Throw an error if nbytes is called on a sparse tensor."


Differential Revision: [D20146388](https://our.internmc.facebook.com/intern/diff/D20146388)

[ghstack-poisoned]
